8284300: Explicitly reject layouts that don't have the correct alignment during linking

JornVernee opened the pull request:
I realized that rejecting layouts that don't have natural alignment also prevents us from supporting packed structs in the future. I tested out the packed struct support and it works on Windows pretty much out of the box. It doesn't work on SysV, some changes to CallArranger are needed. So, there are some options:
1. reject layouts that don't have natural alignment on all platforms (Windows won't be able to use packed structs)
2. reject layouts that don't have natural alignment on platforms besides Windows (yay Windows users)
3. align the status quos: Windows and AArch64 re-compute the alignment. It's not used on Windows, so it can be dropped there. On AArch64 we could "trust" the alignment that is attached to the layout given by the user, same as SysV, but it would fail for packed structs. With the intention of fixing the failure in the future.

I'm leaning most towards 1 at the moment, and that's what this patch implements. We can change that approach when we add reliable support for packed structs across platforms.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8284300](https://bugs.openjdk.org/browse/JDK-8284300): Explicitly reject layouts that don't have the correct alignment during linking


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/728/head:pull/728` \
`$ git checkout pull/728`

Update a local copy of the PR: \
`$ git checkout pull/728` \
`$ git pull https://git.openjdk.org/panama-foreign pull/728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 728`

View PR using the GUI difftool: \
`$ git pr show -t 728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/728.diff">https://git.openjdk.org/panama-foreign/pull/728.diff</a>

</details>

JornVernee commented on src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -94,14 +94,6 @@ private SharedUtils() {
-     * Align the specified type from a given address
-     * @return The address the data should be at based on alignment requirement
-     */
-    public static long align(MemoryLayout t, boolean isVar, long addr) {
```
Notice this code was dead, so I removed it.

mcimadamore commented:
I agree with choice #1. Unless some well defined strategy is defined in some ABI, it doesn't seem like a slam dunk.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/x64/sysv/TypeClass.java:
```diff
@@ -174,6 +176,7 @@ private static List<ArgumentClassImpl> classifyStructType(GroupLayout type) {
     }
 
     static TypeClass classifyLayout(MemoryLayout type) {
+        SharedUtils.checkHasNaturalAlignment(type);
```
Isn't this check too deep in the call chain? E.g. isn't this some validation we could do immediately in the linker, by looking at the function descriptor being passed? Does it need to live in platform dependent code?

JornVernee replied:
Good point, it should be moved up. I went through an iteration where I only checked ValueLayouts, for which it made more sense to put here.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -375,6 +367,21 @@ public static NoSuchElementException newVaListNSEE(MemoryLayout layout) {
+    }
+
+    private static long naturalByteAlignment(MemoryLayout layout) {
+        return alignment(layout, true);
```
Can you please take a look at that `alignment` routine? I'm not 100% sure it's correct (e.g. for arrays it returns 16 - is that true on all platforms?). Also, note that `AbstractLayout` has an `hasNaturalAlignment` predicate, which seems close to what you need.

JornVernee replied:
Hmm yeah, IIRC the array alignment is only 16 if it is a global variable, so I think we can drop that here in this case. I'll use `hasNaturalAlignment` instead. Good suggestion!

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:
```diff
@@ -264,24 +263,27 @@ protected BindingCalculator(boolean forArguments) {
-                if (offset + STACK_SLOT_SIZE < layout.byteSize()) {
+                long copy = Math.min(layout.byteSize() - offset, COPY_CHUNK_SIZE);
+                VMStorage storage = storageCalculator.stackAlloc(copy, align);
+                align = 1; // only first copy needs aligning
```
so, this was an issue in the old code?

JornVernee replied:
Sorry, forgot to explain this.

In the old code we always passed STACK_SLOT_SIZE. This works out in the current implementation since we only allow values to be copied to the stack at stack slot aligned offset. Though, technically, we might need a more precise alignment in the case of Mac/AArch64. For instance, if a struct has 4 byte alignment, we might need to spill it at a 4-byte aligned stack offset.

I noticed this when looking through the code and decided to fix this. But, I'm think now that it kinda doesn't belong in this patch. I can remove it. All we need is the change to drop `SharedUtils.alignment` for `layout.byteAlignment` above.

JornVernee replied:
I think this might have slipped over from when I was working on the Mac/AArch64 stack spill issue, and I didn't think to remove it :/

JornVernee commented:
I've addressed the review comments. I've reverted the changes to the aarch64 CallArranger, since they really don't do anything at the moment.

I also switched to using `AbstractLayout::hasNaturalAlignment` called from AbstractLinker. For aggregates we also have to recursively check the member layouts.

The old alignment computing code was no longer used, so I removed.

I also had a look at the layouts clang gives us for the different platforms, and I don't see any 16 byte alignments for arrays larger than 16 bytes or incomplete arrays, so I think we don't need to take that into account.

JornVernee commented on src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:
```diff
@@ -77,4 +85,29 @@ protected abstract MemorySegment arrangeUpcall(MethodHandle target, MethodType t
+        checkHasNaturalAlignment(layout);
+        if (layout instanceof GroupLayout gl) {
+            for (MemoryLayout member : gl.memberLayouts()) {
+                checkHasNaturalAlignment(member);
```
Ah, good catch

JornVernee commented:
Should be good now. I've amended the test as well to check the recursive case. Sorry for the messy patch.

mcimadamore reviewed (approved):
Looks good!
